pkg/client/node.go

```diff
@@ -184,22 +185,51 @@ func NodeAddToCluster(ctx context.Context, runtime runtimes.Runtime, node *k3d.N
 	l.Log().Tracef("Resulting node %+v", node)
 
 	k3sURLEnvFound := false
+	k3sURLEnvIndex := -1
 	k3sTokenEnvFoundIndex := -1
 	for index, envVar := range node.Env {
 		if strings.HasPrefix(envVar, k3s.EnvClusterConnectURL) {
 			k3sURLEnvFound = true
+			k3sURLEnvIndex = index
 		}
 		if strings.HasPrefix(envVar, k3s.EnvClusterToken) {
 			k3sTokenEnvFoundIndex = index
 		}
 	}
-	if !k3sURLEnvFound {
-		if url, ok := node.RuntimeLabels[k3d.LabelClusterURL]; ok {
-			node.Env = append(node.Env, fmt.Sprintf("%s=%s", k3s.EnvClusterConnectURL, url))
+
+	if checkK3SURLIsActive(cluster.Nodes, node.Env[k3sURLEnvIndex]) {
+		if !k3sURLEnvFound {
+			if url, ok := node.RuntimeLabels[k3d.LabelClusterURL]; ok {
+				node.Env = append(node.Env, fmt.Sprintf("%s=%s", k3s.EnvClusterConnectURL, url))
+			} else {
+				l.Log().Warnln("Failed to find K3S_URL value!")
+			}
+		}
+	} else {
+		// Use LB if available as registration url for nodes
+		// otherwise fallback to server node
+		var registrationNode *k3d.Node
+		serverLBNodes := util.FilterNodesByRole(cluster.Nodes, k3d.LoadBalancerRole)
+		if len(serverLBNodes) > 0 {
+			registrationNode = serverLBNodes[0]
 		} else {
-			l.Log().Warnln("Failed to find K3S_URL value!")
+			for _, existingNode := range cluster.Nodes {
+				if existingNode.Role == k3d.ServerRole {
+					registrationNode = existingNode
+					break
+				}
+			}
+		}
+
+		if !k3sURLEnvFound {
+			node.Env = append(node.Env, fmt.Sprintf("%s=https://%s:%s", k3s.EnvClusterConnectURL, registrationNode.Name, k3d.DefaultAPIPort))
+		} else {
+			if !checkK3SURLIsActive(cluster.Nodes, node.Env[k3sURLEnvIndex]) {
+				node.Env[k3sURLEnvIndex] = fmt.Sprintf("%s=https://%s:%s", k3s.EnvClusterConnectURL, registrationNode.Name, k3d.DefaultAPIPort)
+			}
```

Comment:
@iwilltry42 I have updated my approach with what you suggested. Please have another look.

Reply:
Fine for me, thank you!